### DISCUSSION TITLE
PAGSurface.nativeRelease出现ANR，这个方法里做了啥耗时操作么

### DISCUSSION THREAD

**Discussion by shijianan201** [Tencent/libpag#2095] [General] [answered]
使用的pag的flutter库 最后定位到原生是PAGSurface.nativeRelease这个方法出现了ANR

**Comment by shijianan201** [comment]
这个方法能放在子线程处理么

**Reply by kevingpqi123** [reply, marked as answer]
这里的方法本身并不耗时， pag 内部是线程安全的，渲染过程中的 flush 方法根据素材的复杂度不同耗时不等，在中低端机种耗时的是这个方法，如果正在执行 flush 方法，PAGSurface.nativeRelease中的方法需要等待 flush 方法执行完才会执行，尽管 flush 方法是异步执行的。
PAGSurface.nativeRelease 中的方法不能放在子线程中，pag 渲染方法的执行依赖于 surface，如果 surface 提前被销毁，可能会产生渲染异常

优化思路：
1、最直接的方法是定位出卡顿的素材，优化素材，中低端机尽量使用纯矢量的素材，相对于含有 BMP 预合成的素材，性能会好很多，或者做一个分机型下发素材的逻辑，中低端机尽量使用简化素材；
2、如果提前知道页面销毁，提前执行终止动画的播放，而不是放在生命周期的最后；
3、如果不涉及全屏播放、文本编辑和占位图替换，可以切换使用 PAGImageView， PAGImageView 支持将渲染结果缓存至本地，可以大幅降低这种问题的概率
